File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/bruteforce/Problem15649.java
package com.keunwon.algorithm.bruteforce;

import java.util.Scanner;

public class Problem15649 {

    public static int[] arr;
    public static boolean[] visited;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        final int n = sc.nextInt();
        final int m = sc.nextInt();

        arr = new int[m];
        visited = new boolean[n];

        dfs(n, m, 0);
    }

    public static void dfs(int n, int m, int depth) {
        if (m == depth) {
            for (int i : arr) {
                System.out.print(i + " ");
            }
            System.out.println();
            return;
        }

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                visited[i] = true;
                arr[depth] = i + 1;
                dfs(n, m, depth + 1);
                visited[i] = false;
            }
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem15649.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: N과 M (1)
 * Level: 실버-3
 **/
class Problem15649 {
    private val answer = mutableListOf<String>()

    fun solution(n: Int, m: Int): List<String> {
        dfs(0, n, IntArray(m), BooleanArray(10))
        return answer.sorted()
    }

    private fun dfs(depth: Int, n: Int, picks: IntArray, visited: BooleanArray) {
        if (depth == picks.size) {
            answer.add(picks.joinToString(" "))
            return
        }

        for (i in 1..n) {
            if (visited[i]) continue

            visited[i] = true
            picks[depth] = i
            dfs(depth + 1, n, picks, visited)
            visited[i] = false
        }
    }
}

fun main() {
    val (n, m) = readLine()!!.split(" ").map { it.toInt() }
    Problem15649().solution(n, m).forEach(::println)
}
